Life/Vision/mob hud refactor

phil235 opened the pull request:
This PR does 3 things mostly:
- It removes many updates in life() every tick, such as mob sight and vision overlay, mob health, mob status effects (outside of the natural decrease of certain mob vars like stunned). These updates are now done only when needed, i.e. when something changes your eye_blind var we update your vision. This means there are no more delay for visual updates. e.g. the visual overlay of the welding helmet is added/removed instantaneously when equipped/unequipped, switching from one camera to another when using the security camera console is instantaneous.
- It refactors the mob's datum/hud code to be less shitty. A mob's hud is no longer deleted and recreated every time the player logs in. The ability to toggle your hud with f12 is now available to all mobs.
- It adds the possibility to change the mob's sight and vision when inside something or looking through a camera. The current uses are: your vision is impaired when inside a closet, disposal bin, morgue container, if you have xray vision and use the camera computer looking through a camera you do not see through walls,  when ventcrawling in gaspipes you do not see objects or mobs outside the pipes (unless you have thermal/xray vision).

:cl: phil235
tweak: The vision updates for mobs are now instantaneous (e.g. you don't have to wait one second to see again when removing your welding helmet)
tweak: Your vision is now affected when you're inside something or viewing through a camera. You can no longer see mobs or objects when ventcrawling (unless you have xray for example), an xray user viewing through a camera do not see through walls around the camera. On the other hand, cameras with an xray upgrade let you see through walls. Unfocused camera now give you the same effect as when you are nearsighted. Being inside closets, morgue container and disposal bins give you some vision impairments.
rscadd: The ability to toggle your hud on and off (with f12) is now available to all mobs not just humans.
/:cl:

Other things changed:
- you now only see turfs when inside gas pipes.
- I renamed some vars of datum/hud to be more selfexplanatory
- Moved all datum/hud mob code into the hud folder.
- fixed alien's zone selection button not using the correct sprites.
- I removed the update_hud() proc (that needed to be removed).
- Fixed a typo in /mob/living/carbon/ContractDisease , using "internals" instead of "internal" (very different things)
- Fixed doTeleport() calling Entered() twice on the destination area.
- To reference a mob's selected zone, you now use a direct mob var ("H.zone_selected" instead of "H.zone_sel.selecting")
- mobs lose certain screen objects var ("healths", "zone_sel", "internals", etc) which are now vars of the mob's datum/hud instead.
- the Blind spell is now done via the blind mutation instead of the blind disabilities.
- Give to mobs a version of forceMove(), so the mob is always properly unbuckled, his pull stopped, his vision updated, etc.
- The "user" var of mob/camera/aiEye/remote is renamed to "eye_user" to avoid confusion.
- reset_view() is replaced by reset_perspective(). Now all changes to client.eye and client.perspective are done with this proc.
- I reworked /obj/machinery/computer/security code, changing camera is instantaneous now, as well as cancelling.
- I reworked /obj/machinery/computer/camera_advanced code as well.
- I changed /obj/item/mecha_parts/mecha_equipment/mining_scanner's meson view to be constant instead of by intermittent.
- Fixes not being able to use /obj/item/device/camera_bug while buckled.
- removed admin_forcemove() proc, admin force moving now uses forceMove() simply.
- Removed the client var "adminobs"
- Added var/vision_correction to glasses.
- Added a thermal_overload() proc for glasses, to remove copypasta in emp_act code.
- Remove the hal_crit mob var
- We no longer delete the mob's hud everytime he logs in.
- Being inside disposal bin lowers your vision, like wearing a welding helmet.
- removed the remote_view mob var.
- I changed advanced camera EYE, some fixes, removed unnecessary code when the eye moves, now the mob client eye properly follows the camera mob.
- fixes mob var "machine" not being nullified on logout.
- larva/death() was calling two "living_mob_list -= src"
- I made the Blind screen objects into a global_hud instead of giving one to each mob (like damage overlay).
- I untied tint and eye_blind, TINT_BLIND doesn't give you eye_blind=1.
- gave a visual overlay when inside locker (vimpaired)
- when inside gas pipes you get sight |= (BLIND|SEE_TURFS)
- The new adjust procs serve to properly change eye_blind etc and call vision update procs when needed.
- I added an on_unset_machine() proc to handle perspective reset for camera consoles.
- I moved consequences of eye_check fail inside eye_check() procs themselves.
- I fixed vision updates being fucked by forceMove, especially pipe vision.
- I decided that damage overlay not appearing when dead.
- mob's hud_used is no longer deleted on each login()
- I refactored mob huds a bit, creating subtypes for each mob (/datum/hud/human)
- f12's hud toggling is now available to all mobs
- gave borgs a low_power_mode var so unpowered borg do not use stat= UNCONSCIOUS (which made things weird since you were unconscious but not blind)
- Fixed double Area entering when forced teleporting.
- I fixed larva pulling not being broken when cuffing them, and larva not seeing handcuff alert (and they can resist by clicking it)
- I removed pull updates from life() since it onyl checked for puller's incapacitation.
- I renamed camera/deactivate() to toggle_cam() to be more accurate.
- I fixed mmi brain being immortal (by removing the brain and putting it back)
- I simplified mmi brain emp damage.
- Removed the now unused mob var "permanent_sight_flags"

phil235 committed:
- you now only see turfs when inside gas pipes.
- I renamed some vars of datum/hud to be more selfexplanatory
- Moved all datum/hud mob code into the hud folder.
- fixed alien's zone selection button not using the correct sprites.
- I removed the update_hud() proc (that needed to be removed).
- Fixed a typo in /mob/living/carbon/ContractDisease , using "internals" instead of "internal" (very different things)
- Fixed doTeleport() calling Entered() twice on the destination area.
- To reference a mob's selected zone, you now use a direct mob var ("H.zone_selected" instead of "H.zone_sel.selecting")
- mobs lose certain screen objects var ("healths", "zone_sel", "internals", etc) which are now vars of the mob's datum/hud instead.
- the Blind spell is now done via the blind mutation instead of the blind disabilities.
- Give to mobs a version of forceMove(), so the mob is always properly unbuckled, his pull stopped, his vision updated, etc.
- The "user" var of mob/camera/aiEye/remote is renamed to "eye_user" to avoid confusion.
- reset_view() is replaced by reset_perspective(). Now all changes to client.eye and client.perspective are done with this proc.
- I reworked /obj/machinery/computer/security code, changing camera is instantaneous now, as well as cancelling.
- I reworked /obj/machinery/computer/camera_advanced code as well.
- I changed /obj/item/mecha_parts/mecha_equipment/mining_scanner's meson view to be constant instead of by intermittent.
- Fixes not being able to use /obj/item/device/camera_bug while buckled.
- removed admin_forcemove() proc, admin force moving now uses forceMove() simply.
- Removed the client var "adminobs"
- Added var/vision_correction to glasses.
- Added a thermal_overload() proc for glasses, to remove copypasta in emp_act code.
- Remove the hal_crit mob var
- We no longer delete the mob's hud everytime he logs in.
- Added a stat == dead check in mob's metabolize() so we immediately stop metabolizing if one of the chem kills the mob.
- Being inside disposal bin lowers your vision, like wearing a welding helmet.
- removed the remote_view mob var.
- I changed advanced camera EYE, some fixes, removed unnecessary code when the eye moves, now the mob client eye properly follows the camera mob.
- fixes mob var "machine" not being nullified on logout.
- larva/death() was calling two "living_mob_list -= src"
- I made the Blind screen objects into a global_hud instead of giving one to each mob (like damage overlay).
- I untied tint and eye_blind, TINT_BLIND doesn't give you eye_blind=1.
- gave a visual overlay when inside locker (vimpaired)
- when inside disposal/gas pipes you get sight |= (BLIND|SEE_TURFS)
- glasses toggling updates (atmos meson toggle): DONE
- The new adjust procs serve to properly change eye_blind etc and call vision update procs when needed.
- I added an on_unset_machine() proc to handle perspective reset for camera consoles.
- I moved consequences of eye_check fail inside eye_check() procs themselves.
- I fixed vision updates being fucked by forceMove, especially pipe vision.
- I decided that damage overlay not appearing when dead.
- mob's hud_used is no longer deleted on each login()
- I refactored mob huds a bit, creating subtypes for each mob (/datum/hud/human)
- f12's hud toggling is now available to all mobs
- gave borgs a low_power_mode var so unpowered borg do not use stat= UNCONSCIOUS (which made things weird since you were unconscious but not blind)
- Fixed double Area entering when forced teleporting.
- I fixed larva pulling not being broken when cuffing them, and larva not seeing handcuff alert (and they can resist by clicking it)
- I removed pull updates from life() since it onyl checked for puller's incapacitation.
- I renamed camera/deactivate() to toggle_cam() to be more accurate.
- I fixed mmi brain being immortal (by removing the brain and putting it back)
- I simplified mmi brain emp damage.

neersighted commented:
Wew! This looks exciting.

KorPhaeron commented:
>you now only see turfs when inside gas pipes.

Did trying to fix that bug lead to an entire rewrite?

ChangelingRain commented on code/_onclick/hud/blob_overmind.dm:
```diff
@@ -0,0 +1,151 @@
+/obj/screen/blob/Blobbernaut
+	icon_state = "ui_blobbernaut"
+	name = "Produce Blobbernaut (20)"
+	desc = "Produces a blobbernaut for 20 points."
```
It's 30 points now.
Is this comment useful? Maybe.

phil235 commented:
@KorPhaeron No, actually I just realized I could fix that bug through my new update_remote_sight() proc.

@ChangelingRain the PR is on a month old branch, I have yet to fix the merge conflicts.

PKPenguin321 commented:
holy shit phil are you a god

Cebasiz commented:
>this PR does three things mostly
>lists a thousand fucking cool things

thanks

Lumipharon commented:
Jesus fuck all these good fixes.

KorPhaeron commented:
>gave a visual overlay when inside locker (vimpaired)

Is this welding vision? Don't really think hiding/stealth needs to be discouraged personally and I wish balance changes weren't part of the refactor, but I guess welder vision is tolerable.

phil235 commented:
@KorPhaeron No it's the same impairement you get when nearsighted, less obstruction than welding helmet.

bgobandit commented:
does this close any issues, it has to right

RemieRichards commented on code/game/gamemodes/changeling/powers/horror_form.dm:
```diff
@@ -0,0 +1,28 @@
+/obj/effect/proc_holder/changeling/horror_form //Horror Form: turns the changeling into a terrifying abomination
```
Umm.

PKPenguin321 replied:
spooky

phil235 replied:
woops, I didn't mean to add this file.

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into VisionUpdateRefactor

Conflicts:
	code/ATMOSPHERICS/components/unary_devices/cryo.dm
	code/_onclick/hud/alert.dm
	code/_onclick/hud/hud.dm
	code/datums/mutations.dm
	code/datums/wires/robot.dm
	code/game/atoms.dm
	code/game/gamemodes/blob/overmind.dm
	code/game/machinery/alarm.dm
	code/game/machinery/machinery.dm
	code/game/machinery/suit_storage_unit.dm
	code/game/objects/items/weapons/tanks/tanks.dm
	code/game/objects/items/weapons/tools.dm
	code/game/objects/structures/morgue.dm
	code/modules/admin/verbs/adminjump.dm
	code/modules/atmospherics/machinery/atmosmachinery.dm
	code/modules/mob/inventory.dm
	code/modules/mob/living/carbon/alien/humanoid/death.dm
	code/modules/mob/living/carbon/alien/larva/death.dm
	code/modules/mob/living/carbon/brain/death.dm
	code/modules/mob/living/carbon/carbon.dm
	code/modules/mob/living/carbon/human/death.dm
	code/modules/mob/living/carbon/human/human.dm
	code/modules/mob/living/carbon/human/human_damage.dm
	code/modules/mob/living/carbon/human/life.dm
	code/modules/mob/living/carbon/human/species.dm
	code/modules/mob/living/carbon/human/species_types.dm
	code/modules/mob/living/carbon/life.dm
	code/modules/mob/living/carbon/monkey/death.dm
	code/modules/mob/living/life.dm
	code/modules/mob/living/living.dm
	code/modules/mob/living/silicon/ai/ai.dm
	code/modules/mob/living/silicon/ai/death.dm
	code/modules/mob/living/silicon/ai/life.dm
	code/modules/mob/living/silicon/pai/death.dm
	code/modules/mob/living/silicon/pai/pai.dm
	code/modules/mob/living/silicon/robot/death.dm
	code/modules/mob/living/silicon/robot/life.dm
	code/modules/mob/living/silicon/robot/robot.dm
	code/modules/mob/living/silicon/silicon.dm
	code/modules/mob/living/simple_animal/guardian/guardian.dm
	code/modules/mob/login.dm
	code/modules/mob/mob.dm
	code/modules/projectiles/gun.dm
	code/modules/reagents/chemistry/reagents/blob_reagents.dm
	tgstation.dme

phil235 committed:
More merge conflict fixes.

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into VisionUpdateRefactor

phil235 committed:
Fixes merge conflict erro (readd datum/hud/Destroy())
Fixes cyborg sight not updating when equiping/unequiping sight modules.

phil235 committed:
Some errors and merge conflicts fixing mistakes.

phil235 commented:
The PR is ready, I'm going to wait for #14981 to be merged, and I'll fix conflicts.

Dorsisdwarf commented:
Oh boy time to track down all the issues this fixes

Dorsisdwarf commented:
Fixes #7033.

phil235 commented:
@Dorsisdwarf I doesn't fix 7033 entirely, you still see mobs while inside DISPOSAL pipes. I might try to fix it in a future PR.

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into VisionUpdateRefactor

Conflicts:
	code/_onclick/hud/alien.dm
	code/_onclick/hud/alien_larva.dm
	code/_onclick/hud/hud.dm
	code/_onclick/hud/human.dm
	code/_onclick/hud/monkey.dm
	code/_onclick/hud/other_mobs.dm
	code/_onclick/hud/robot.dm
	code/game/machinery/Sleeper.dm
	code/modules/mob/living/carbon/alien/larva/death.dm
	code/modules/mob/living/carbon/brain/death.dm
	code/modules/mob/living/carbon/human/death.dm
	code/modules/mob/living/carbon/human/life.dm
	code/modules/mob/living/carbon/human/species.dm
	code/modules/mob/living/carbon/life.dm
	code/modules/mob/living/carbon/monkey/death.dm
	code/modules/mob/living/death.dm
	code/modules/mob/living/life.dm
	code/modules/mob/living/silicon/ai/death.dm
	code/modules/mob/living/silicon/ai/life.dm
	code/modules/mob/living/silicon/ai/login.dm
	code/modules/mob/living/silicon/pai/death.dm
	code/modules/mob/living/silicon/robot/death.dm
	code/modules/mob/living/silicon/robot/robot.dm
	code/modules/mob/mob_defines.dm

phil235 committed:
Made changes to adjust with neersighted's fullscreen overlays pr.
Fixed Xray users getting vision impairment overlays when inside a container.
Fixes mob inside mecha getting the mech sight flags despite not being the pilot.
"get_vision_impairments" is renamed to "get_remote_view_fullscreens".
Fixes AI blindness.

phil235 committed:
Fixes fullscreen overlay not appearing on login.

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into VisionUpdateRefactor

Conflicts:
	code/modules/reagents/chemistry/holder.dm

phil235 commented:
Alright, it should ready now.

phil235 committed:
Removing the true_changeling.dm file that I added by mistake.
renamed "nearsight" fullscreen to "nearsighted"
